torch/csrc/generic/Tensor.cpp

```diff
@@ -47,32 +77,61 @@
   THHostStorage_(free)(cpu_storage);                           \
 }
 
-#define COPY_FROM_ARRAY_CUDA_HALF(ELTYPE, ARRAY, STORAGE, SIZE) \
+#define COPY_FROM_HALF_ARRAY_CUDA(ELTYPE, ARRAY, STORAGE, SIZE) \
+{ \
+  ELTYPE *arrdata = (ELTYPE*)PyArray_DATA(ARRAY);               \
+  std::unique_ptr<load_real> data_guard(new load_real[SIZE]);   \
+  load_real *data = data_guard.get();                           \
+  for (size_t i=0; i<SIZE; i++) {                               \
+    data[i] = TH_half2float(arrdata[i]);                        \
+  }                                                             \
+  THHostStorage *cpu_storage =                                  \
+      THHostStorage_(newWithData)(data_guard.get(), SIZE);      \
+  cpu_storage->flag &= ~TH_STORAGE_FREEMEM;                     \
+  THCStorage_(copyCPU)(LIBRARY_STATE STORAGE, cpu_storage);     \
+  THHostStorage_(free)(cpu_storage);                            \
+}
```

Comment:
idea: add an extra argument to `COPY_FROM_ARRAY_CUDA`, which is a caster for `arrdata`. By default it's identity, for half it adds `TH_half2float`.